onnxruntime/core/framework/device_stream_collection.h

```diff
@@ -0,0 +1,37 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+#pragma once
+#include "core/framework/stream_handles.h"
+#include "gsl/gsl"
+
+namespace onnxruntime {
+class SessionState;
+
+class DeviceStreamCollectionImpl;
+// a collection of indexed device stream instances.
+// the index is the same as the logic sequence index in the execution plan.
+// this collection may be cached and reused for future iterations.
+class DeviceStreamCollection {
```

Comment:
From design perspective, I think we should provide an API to get specific stream from index:

Stream* GetStream(size_t index);

And remove GetStreams() if possible. This is to hide and not expose the implementation details (i.e., gsl::span<Stream*> objects) from the callers. By doing this, DeviceStreamCollection is the only object managing gsl::span<Streams*>